### PR TITLE
Cloning (v2 2): Electric Boogaloo

### LINKED ISSUE #20: Cloning? Only if it can be made 100% modular (Skyrat had it for some time, maybe you can look there).

### PR DESCRIPTION
**Pull request by PeriodicChaos**
## About The Pull Request

_**Cloning (v2): Modular Edition**_  Added back cloning from [monke](https://github.com/Monkestation/MonkeStation/pull/213/files#diff-11e7b481591c36107bd40e6c36d5838f24d9a3979d8e0a3a373b5e45a2679011)!! _(Bee Station Cloning)_

## Why It's Good For The Game

Its a helpful tool to come back to life and not wait for those pesky crewmates to add more members in the **_Cryopods_** and wait for a **_5-minute timer_** to be up to even start respawning.  All around win the players.

## Notes
Right now cloning works and some Unused/Broken Features are gone I don't think we need much more work done on cloning we can even add it in now if we like the rest can be edited later on all that matters is it works.

## To Do List
- [x] **Make Cloning Code Green** - _The code is approved by the coding gods to run_
- [ ] **Make Cloning Work Ingame** - _When I review the code and test it ingame_
- [ ] **Get Merge Ready**- _This will be a bit_
cloning
## Changelog
:cl:
add: Cloning (v2): Modular Edition Added Cloning back have fun not waiting for the respawn timers!
/:cl:

Closes #20

### CODE REVIEW

**Comment by ZephyrTFA** [comment]
cloning 2 2: electric Boogaloo

**Comment by PeriodicChaos** [comment]
> cloning 2 2: electric Boogaloo

Your right that is its proper name now

File: voidcrew/modules/cloning/code/computer_cloning [hunk 1]
```diff
@@ -0,0 +1,630 @@
+#define AUTOCLONING_MINIMAL_LEVEL 3
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/computer_cloning:1]: This isn't a .dm file

Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/computer_cloning:1]: will fix that no idea why it reverted back from being one

File: voidcrew/_DEFINES/cloning_defines.dm [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/// New or not new clone
+#define CLONER_FRESH_CLONE "fresh"
```
Comment (JohnFulpWillard) [inline voidcrew/_DEFINES/cloning_defines.dm:2]: This define is never used. We check for it, but it's never set by anything ever. We probably don't need the other define too, or the var that uses it. Just hardcode it to work as if it's always using the second define.

File: voidcrew/modules/cloning/code/cloning.dm [hunk 3]
```diff
@@ -0,0 +1,589 @@
+//Cloning revival method.
+//The pod handles the actual cloning while the computer manages the clone profiles
+
+//Potential replacement for genetics revives or something I dunno (?)
+
+#define CLONE_INITIAL_DAMAGE     150    //Clones in clonepods start with 150 cloneloss damage and 150 brainloss damage, thats just logical
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:6]:
```suggestion
///Clones in clonepods start with 150 cloneloss damage and 150 brainloss damage, thats just logical
#define CLONE_INITIAL_DAMAGE 150
```

File: voidcrew/modules/cloning/code/cloning.dm [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+	var/mob/living/mob_occupant = occupant
+	go_out()
+	if(mob_occupant)
+		log_combat("[key_name(mob_occupant)] ejected from [src] at [AREACOORD(src)] due to Destroy().")
+	QDEL_NULL(radio)
+	QDEL_NULL(countdown)
+	if(connected)
+		connected.DetachCloner(src)
+	QDEL_LIST(unattached_flesh)
+	. = ..()
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:67]:
```suggestion
	return ..()
```

File: voidcrew/modules/cloning/code/cloning.dm [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+		radio.subspace_transmission = TRUE
+		radio.canhear_range = 0
+		radio.recalculateChannels()
+	update_icon()
+
+/obj/machinery/clonepod/Destroy()
+	var/mob/living/mob_occupant = occupant
+	go_out()
+	if(mob_occupant)
+		log_combat("[key_name(mob_occupant)] ejected from [src] at [AREACOORD(src)] due to Destroy().")
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:61]:
```suggestion
	go_out()
	if(occupant)
		log_combat("[key_name(occupant)] ejected from [src] at [AREACOORD(src)] due to Destroy().")
```

File: voidcrew/modules/cloning/code/objects_datums.dm [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+/datum/brain_trauma/special/beepsky
+	clonable = FALSE
+
+/datum/mood_event/cloned_corpse
+	description = "<span class='boldwarning'>I recently saw my own corpse...</span>\n"
+	mood_change = -6
+
+/datum/quirk/proc/clone_data() //return additional data that should be remembered by cloning
+/datum/quirk/proc/on_clone(data) //create the quirk from clone data
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/objects_datums.dm:40]: I don't think we need to do this at all

File: voidcrew/modules/cloning/code/exp_cloner.dm [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+/obj/machinery/clonepod/experimental/growclone(clonename, ui, mutation_index, mindref, last_death, blood_type, datum/species/mrace, list/features, list/mutantparts, list/markings, factions, list/quirks, datum/bank_account/insurance, list/traumas, empty)
+	if(panel_open)
+		return NONE
+	if(mess || attempting)
+		return NONE
+
+	attempting = TRUE //One at a time!!
+	countdown.start()
+
+	var/mob/living/carbon/human/H = new /mob/living/carbon/human(src)
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/exp_cloner.dm:21]: Do you think we can just make use of ``duplicate_object`` helper here?

File: voidcrew/modules/cloning/code/exp_cloner.dm [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+		to_chat(H.get_ghost(TRUE), "<span class='notice'>Your body is beginning to regenerate in a cloning pod. You will become conscious when it is complete.</span>")
+
+	if(H)
+		H.faction |= factions
+
+		H.set_cloned_appearance()
+
+		H.set_suicide(FALSE)
+	attempting = FALSE
+	update_icon()
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/exp_cloner.dm:72]:
```suggestion
	update_appearance(UPDATE_ICON)
```

File: voidcrew/modules/cloning/code/exp_cloner.dm [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+	icon_screen = "dna"
+	icon_keyboard = "med_key"
+	circuit = /obj/item/circuitboard/computer/prototype_cloning
+	var/obj/machinery/dna_scannernew/scanner = null //Linked scanner. For scanning.
+	var/list/pods //Linked experimental cloning pods
+	var/temp = "Inactive"
+	var/scantemp = "Ready to Scan"
+	var/loading = FALSE // Nice loading text
+
+	light_color = LIGHT_COLOR_BLUE
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/exp_cloner.dm:89]: above var defines please

File: voidcrew/modules/cloning/code/exp_cloner.dm [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+/obj/machinery/computer/prototype_cloning/interact(mob/user)
+	user.set_machine(src)
+	add_fingerprint(user)
+
+	if(..())
+		return
+
+	updatemodules(TRUE)
+
+	var/dat = ""
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/exp_cloner.dm:181]: do we not have a TGUI page for this?

**Review by JohnFulpWillard — CHANGES_REQUESTED** [review]

File: voidcrew/modules/cloning/code/cloning_boards.dm [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+		/obj/item/stack/sheet/glass = 1)
+
+/obj/item/circuitboard/machine/clonepod/experimental
+	name = "Experimental Clone Pod"
+	greyscale_colors = CIRCUIT_COLOR_MEDICAL
+	build_path = /obj/machinery/clonepod/experimental
+
+//Machine Designs
+
+/datum/design/board/clonecontrol
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning_boards.dm:32]: Just updated your branch to master, please move it to the rest of the designs as shown in https://github.com/Willardstation/tg-voidcrew/pull/79

File: voidcrew/modules/cloning/code/objects_datums.dm [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+//techweb
+/datum/techweb_node/cloning
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/objects_datums.dm:43]: Please move this to the other techweb nodes as shown in https://github.com/Willardstation/tg-voidcrew/pull/79

**Comment by PeriodicChaos** [comment]
Sorry about the code change up forgot that the code that was here before was older the new stuff is here now

File: voidcrew/modules/cloning/code/objects_datums.dm (hunk 12 above)
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/objects_datums.dm:47]: will do

File: voidcrew/modules/cloning/code/cloning_boards.dm (hunk 11 above)
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/cloning_boards.dm:33]: will do

File: voidcrew/_DEFINES/cloning_defines.dm (hunk 2 above)
Comment (PeriodicChaos) [reply voidcrew/_DEFINES/cloning_defines.dm:2]: I will check it out and do that

File: voidcrew/modules/cloning/code/cloning.dm [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+/obj/machinery/clonepod/handle_atom_del(atom/deleting_atom)
+	if(deleting_atom == occupant)
+		occupant = null
+		countdown.stop()
+
+/obj/machinery/clonepod/proc/horrifyingsound()
+	for(var/i in 1 to 5)
+		playsound(src, pick('sound/hallucinations/growl1.ogg','sound/hallucinations/growl2.ogg','sound/hallucinations/growl3.ogg'), 100, rand(0.95, 1.05))
+		sleep(1)
+	sleep(10)
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:538]: Please use timers instead of sleep

File: voidcrew/modules/cloning/code/cloning.dm [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+
+/obj/machinery/clonepod/examine(mob/user)
+	. = ..()
+	. += span_notice("The <i>linking</i> device can be <i>scanned<i> with a multitool. It can be emptied by Alt-Clicking it.")
+	if(in_range(user, src) || isobserver(user))
+		. += span_notice("The status display reads: Cloning speed at <b>[speed_coeff*50]%</b>.<br>Predicted amount of cellular damage: <b>[100-heal_level]%</b><br> Storing up to <b>[reagents.maximum_volume]cm<sup>3</sup></b> of synthflesh.")
+		. += span_notice("Synthflesh consumption at <b>[round(fleshamnt*90, 1)]cm<sup>3</sup></b> per clone.")
+		. += span_notice("The reagent display reads: [round(reagents.total_volume, 1)] / [reagents.maximum_volume] cm<sup>3</sup>")
+		if(efficiency > 5)
+			. += span_notice("Pod has been upgraded to support autoprocessing and apply beneficial mutations.")
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:128]: Do we really want people to get free mutations from killing and autocloning themselves??
I thought we had agreed to remove the feature entrely and only have it be "no negative mutations"

**Review by JohnFulpWillard — CHANGES_REQUESTED** [review]

File: voidcrew/modules/cloning/code/cloning.dm (hunk 14 above)
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/cloning.dm:128]: Yeah i can remove the mutations from happening I think I forgot to remove them

File: voidcrew/modules/cloning/code/computer_cloning.dm [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+	req_access = list(ACCESS_GENETICS) //for modifying records
+	light_color = LIGHT_COLOR_BLUE
+
+	///List of all user records we have stored on the computer.
+	var/list/datum/data/record/records = list()
+
+	///The disk inserted into the console, holding cloning data for patients that we can read/write from/to.
+	var/obj/item/disk/data/diskette
+	///The linked machine that holds the bodies we can scan.
+	var/obj/machinery/dna_scannernew/scanner
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/computer_cloning.dm:22]: ``scanner`` is never used/set by anything.

File: voidcrew/modules/cloning/code/computer_cloning.dm [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+		START_PROCESSING(SSmachines, src)
+
+/obj/machinery/computer/cloning/proc/findscanner()
+	for(var/direction in GLOB.cardinals)
+		var/obj/machinery/dna_scannernew/scannerf = locate(/obj/machinery/dna_scannernew, get_step(src, direction))
+		// If found and operational, return the scanner
+		if (scannerf && scannerf.is_operational)
+			return scannerf
+	// If no scanner was found, it will return null
+	return null
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/computer_cloning.dm:151]: This proc is never called

File: voidcrew/modules/cloning/code/computer_cloning.dm (hunk 15 above)
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/computer_cloning.dm:22]: I think its used by the pod the computer gets connected to the dna_scannernew

File: voidcrew/modules/cloning/code/computer_cloning.dm [hunk 17]
```diff
@@ -0,0 +1,526 @@
...
+		temp = "[mob_occupant.real_name] => Cloning data sent to pod."
+		playsound(src, 'sound/machines/terminal_prompt_confirm.ogg', 50, 0)
+		log_combat("[user ? key_name(user) : "Unknown"] cloned [key_name(mob_occupant)] with [src] at [AREACOORD(src)].")
+
+/obj/machinery/computer/cloning/proc/can_scan(datum/dna/dna, mob/living/mob_occupant, datum/bank_account/account, body_only)
+	if(!istype(dna))
+		scantemp = "Unable to locate valid genetic data."
+		playsound(src, 'sound/machines/terminal_prompt_deny.ogg', 50, 0)
+		return FALSE
+	if(NO_DNA_COPY in dna.species.species_traits)
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/computer_cloning.dm:415]: is this still functional?

**Review by JohnFulpWillard — COMMENTED** [review]
my other reviews please

File: voidcrew/modules/cloning/code/computer_cloning.dm (hunk 17 above)
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/computer_cloning.dm:415]: So It seems i used the wrong code once more... I will go fix up the code this was to use

Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/computer_cloning.dm:415]: I don't think so

File: voidcrew/modules/cloning/code/objects_datums.dm [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+/datum/gas_mixture/immutable/cloner/garbage_collect()
+	..()
+	ADD_GAS(/datum/gas/nitrogen, gases)
+	gases[/datum/gas/nitrogen][MOLES] = MOLES_O2STANDARD + MOLES_N2STANDARD
+
+/datum/gas_mixture/immutable/cloner/heat_capacity()
+	return (MOLES_O2STANDARD + MOLES_N2STANDARD)*20 //specific heat of nitrogen is 20
+
+/datum/data/record
```
Comment (PeriodicChaos) [inline voidcrew/modules/cloning/code/objects_datums.dm:57]: This right here is the place holder for i have no idea how to get this data

**Comment by PeriodicChaos** [comment]
In objects_datum 

/datum/data/record
	name = "record"
	var/list/fields = list()

is the place holder

**Pull request closed without merging**